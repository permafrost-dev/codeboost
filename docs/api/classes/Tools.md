[codeboost](../README.md) / [Exports](../modules.md) / Tools

# Class: Tools

## Table of contents

### Constructors

-   [constructor](Tools.md#constructor)

### Methods

-   [copyfile](Tools.md#copyfile)
-   [exec](Tools.md#exec)
-   [fileexists](Tools.md#fileexists)
-   [filesAreEqual](Tools.md#filesareequal)
-   [hashfile](Tools.md#hashfile)
-   [hashstring](Tools.md#hashstring)
-   [readJson](Tools.md#readjson)
-   [readYaml](Tools.md#readyaml)
-   [readfile](Tools.md#readfile)
-   [recursiveDirectoryCopy](Tools.md#recursivedirectorycopy)
-   [sleep](Tools.md#sleep)
-   [writeJson](Tools.md#writejson)
-   [writeYaml](Tools.md#writeyaml)
-   [writefile](Tools.md#writefile)

## Constructors

### constructor

• **new Tools**()

## Methods

### copyfile

▸ **copyfile**(`src`, `dest`): `void`

Copies a file from `src` to `dest`. Both `src` and `dest` should be absolute paths.

#### Parameters

| Name   | Type     |
| :----- | :------- |
| `src`  | `string` |
| `dest` | `string` |

#### Returns

`void`

---

### exec

▸ **exec**(`command`, `silent?`): `string`

It executes a command and returns the output as a string

#### Parameters

| Name      | Type      | Default value | Description                                         |
| :-------- | :-------- | :------------ | :-------------------------------------------------- |
| `command` | `string`  | `undefined`   | The command to execute.                             |
| `silent?` | `boolean` | `true`        | If true, the command will not be printed to stdout. |

#### Returns

`string`

The output of the command.

---

### fileexists

▸ **fileexists**(`path`): `boolean`

This function returns true if the file exists, and false if it doesn't

#### Parameters

| Name   | Type     | Description                             |
| :----- | :------- | :-------------------------------------- |
| `path` | `string` | The path to the file you want to check. |

#### Returns

`boolean`

A boolean value.

---

### filesAreEqual

▸ **filesAreEqual**(`file1`, `file2`): `boolean`

This function returns true if the two files have the exact same contents

#### Parameters

| Name    | Type     |
| :------ | :------- |
| `file1` | `string` |
| `file2` | `string` |

#### Returns

`boolean`

---

### hashfile

▸ **hashfile**(`path`): `string`

It takes a file path and returns the sha-256 hash of the file contents.

#### Parameters

| Name   | Type     | Description                            |
| :----- | :------- | :------------------------------------- |
| `path` | `string` | The path to the file you want to hash. |

#### Returns

`string`

The sha-256 hash of the file.

---

### hashstring

▸ **hashstring**(`str`): `string`

It takes a string and returns the sha-256 hash.

#### Parameters

| Name  | Type     | Description                  |
| :---- | :------- | :--------------------------- |
| `str` | `string` | The string you want to hash. |

#### Returns

`string`

The sha-256 hash of the string.

---

### readJson

▸ **readJson**(`path`): `any`

It reads a json file and returns the parsed contents as an object.

#### Parameters

| Name   | Type     | Description                            |
| :----- | :------- | :------------------------------------- |
| `path` | `string` | The path to the file you want to read. |

#### Returns

`any`

---

### readYaml

▸ **readYaml**(`path`): `any`

It reads a yaml file and returns the contents as an object.

#### Parameters

| Name   | Type     | Description                            |
| :----- | :------- | :------------------------------------- |
| `path` | `string` | The path to the file you want to read. |

#### Returns

`any`

---

### readfile

▸ **readfile**(`path`): `string`

It reads a file and returns its contents

#### Parameters

| Name   | Type     | Description                            |
| :----- | :------- | :------------------------------------- |
| `path` | `string` | The path to the file you want to read. |

#### Returns

`string`

The contents of the file at the given path.

---

### recursiveDirectoryCopy

▸ **recursiveDirectoryCopy**(`src`, `dest`, `onCopiedCallback?`): `string`[]

A function that recursively copies a files from one directory to another. Files that already exist
at the destination are overwritten, and directories that do not exist are created.

Returns a list of filenames that were copied.

#### Parameters

| Name               | Type                                                                   | Default value | Description                                                   |
| :----------------- | :--------------------------------------------------------------------- | :------------ | :------------------------------------------------------------ |
| `src`              | `string`                                                               | `undefined`   | The source directory.                                         |
| `dest`             | `string`                                                               | `undefined`   | The destination directory.                                    |
| `onCopiedCallback` | `null` \| [`OnFileCopiedCallback`](../modules.md#onfilecopiedcallback) | `null`        | A callback function that is called after each file is copied. |

#### Returns

`string`[]

---

### sleep

▸ **sleep**(`ms`): `Promise`<`unknown`\>

This function takes a number of milliseconds and returns a promise that
resolves after that many milliseconds have passed.

This function is used to create a delay in a program, which can be useful
for debugging, or for building a user interface with an animation.

#### Parameters

| Name | Type     | Description                         |
| :--- | :------- | :---------------------------------- |
| `ms` | `number` | The number of milliseconds to wait. |

#### Returns

`Promise`<`unknown`\>

A promise that resolves after the given number of milliseconds.

---

### writeJson

▸ **writeJson**(`path`, `data`): `void`

It writes a JSON file to the specified path.

#### Parameters

| Name   | Type     | Description                                |
| :----- | :------- | :----------------------------------------- |
| `path` | `string` | The path to the file you want to write to. |
| `data` | `any`    | The content to write to the file.          |

#### Returns

`void`

---

### writeYaml

▸ **writeYaml**(`path`, `data`): `void`

It dumps an object to yaml and saves it to the given path.

#### Parameters

| Name   | Type     | Description                                |
| :----- | :------- | :----------------------------------------- |
| `path` | `string` | The path to the file you want to write to. |
| `data` | `any`    |                                            |

#### Returns

`void`

---

### writefile

▸ **writefile**(`path`, `content`): `void`

Write the given content to the given file path.

#### Parameters

| Name      | Type     | Description                                |
| :-------- | :------- | :----------------------------------------- |
| `path`    | `string` | The path to the file you want to write to. |
| `content` | `string` | The content to write to the file.          |

#### Returns

`void`
