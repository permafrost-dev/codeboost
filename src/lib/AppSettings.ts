/* It initializes a git instance */

import { LogTarget } from '@/traits/HasLogger';

export interface AppSettings {
    github_token: string;
    repository_storage_path: string;
    boosts_path: string;
    use_forks: boolean;
    use_pull_requests: boolean;
    log_target: LogTarget | LogTarget[];
}

export function loadSettings(filename: string): AppSettings {
    const settings = require(filename).default;

    return transformSettings(settings);
}

export function transformSettings(settings: AppSettings): AppSettings {
    // load settings from environment variable references starting with '$'
    for (const key in settings) {
        if (typeof settings[key] === 'string' && settings[key].startsWith('$')) {
            if (typeof process.env[settings[key].slice(1)] !== 'undefined') {
                settings[key] = process.env[settings[key].slice(1)];
            }
        }
    }

    if (!Array.isArray(settings.log_target)) {
        settings.log_target = [settings.log_target];
    }

    settings.log_target = settings.log_target
        .map(target => <LogTarget>target.toLowerCase())
        .filter((target: string) => {
            return ['console', 'file'].includes(target);
        });

    return settings;
}
