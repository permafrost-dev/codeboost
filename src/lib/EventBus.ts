import EventEmitter2 from 'eventemitter2';

export default class EventBus {
    public emitter: EventEmitter2;

    constructor() {
        this.emitter = new EventEmitter2();
    }

    public on(event: string, listener: (...args: any[]) => void) {
        return this.emitter.on(event, listener);
    }

    public off(event: string, listener: (...args: any[]) => void) {
        return this.emitter.off(event, listener);
    }

    public emit(event: string, ...args: any[]) {
        return this.emitter.emit(event, ...args);
    }

    public once(event: string, listener: (...args: any[]) => void) {
        return this.emitter.once(event, listener);
    }
}

export const eventbus = new EventBus();
