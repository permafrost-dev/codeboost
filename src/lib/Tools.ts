import { execSync } from 'child_process';
import { createHash } from 'crypto';
import { copyFileSync, existsSync, lstatSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from 'fs';
import yaml from 'js-yaml';
import { dirname } from 'path';

export type OnFileCopiedCallback = (src: string, dest: string) => void;

export class Tools {
    /**
     * This function takes a number of milliseconds and returns a promise that
     * resolves after that many milliseconds have passed.
     *
     * This function is used to create a delay in a program, which can be useful
     * for debugging, or for building a user interface with an animation.
     * @param {number} ms - The number of milliseconds to wait.
     * @returns A promise that resolves after the given number of milliseconds.
     */
    public async sleep(ms: number) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * This function returns true if the file exists, and false if it doesn't
     * @param {string} path - The path to the file you want to check.
     * @returns A boolean value.
     */
    public fileexists(path: string) {
        return existsSync(path);
    }

    /**
     * It reads a file and returns its contents
     * @param {string} path - The path to the file you want to read.
     * @returns The contents of the file at the given path.
     */
    public readfile(path: string) {
        return readFileSync(path, { encoding: 'utf8' });
    }

    /**
     * Write the given content to the given file path.
     * @param {string} path - The path to the file you want to write to.
     * @param {string} content - The content to write to the file.
     */
    public writefile(path: string, content: string) {
        writeFileSync(path, content, { encoding: 'utf8' });
    }

    /**
     * Copies a file from `src` to `dest`. Both `src` and `dest` should be absolute paths.
     * @param {string} src
     * @param {string} dest
     */
    public copyfile(src: string, dest: string) {
        if (!existsSync(dirname(dest))) {
            mkdirSync(dirname(dest), { recursive: true });
        }

        copyFileSync(src, dest);
    }

    /**
     * It takes a file path and returns the sha-256 hash of the file contents.
     * @param {string} path - The path to the file you want to hash.
     * @returns {string} The sha-256 hash of the file.
     */
    public hashfile(path: string) {
        return this.hashstring(this.readfile(path));
    }

    /**
     * It takes a string and returns the sha-256 hash.
     * @param {string} str - The string you want to hash.
     * @returns {string} The sha-256 hash of the string.
     */
    public hashstring(str: string) {
        return createHash('sha256').update(str)
            .digest('hex')
            .toLowerCase();
    }

    /**
     * This function returns true if the two files have the exact same contents
     * @param {string}
     * @param {string}.
     */
    public filesAreEqual(file1: string, file2: string) {
        if (!this.fileexists(file1) || !this.fileexists(file2)) {
            return false;
        }

        return this.hashfile(file1) === this.hashfile(file2);
    }

    /**
     * It reads a yaml file and returns the contents as an object.
     * @param {string} path - The path to the file you want to read.
     * @returns {object}
     */
    public readYaml(path: string) {
        return yaml.load(this.readfile(path));
    }

    /**
     * It dumps an object to yaml and saves it to the given path.
     * @param {string} path - The path to the file you want to write to.
     * @param {any} data
     */
    public writeYaml(path: string, data: any) {
        this.writefile(path, yaml.dump(data));
    }

    /**
     * It reads a json file and returns the parsed contents as an object.
     * @param {string} path - The path to the file you want to read.
     * @returns {object}
     */
    public readJson(path: string) {
        return JSON.parse(this.readfile(path));
    }

    /**
     * It writes a JSON file to the specified path.
     * @param {string} path - The path to the file you want to write to.
     * @param {any} data - The content to write to the file.
     */
    public writeJson(path: string, data: any) {
        this.writefile(path, JSON.stringify(data, null, 4));
    }

    /**
     * It executes a command and returns the output as a string
     * @param {string} command - The command to execute.
     * @param [silent=true] - If true, the command will not be printed to stdout.
     * @returns {string} The output of the command.
     */
    public exec(command: string, silent = true) {
        return execSync(command, { stdio: silent ? 'pipe' : 'inherit' }).toString();
    }

    /**
     * A function that recursively copies a files from one directory to another. Files that already exist
     * at the destination are overwritten, and directories that do not exist are created.
     *
     * Returns a list of filenames that were copied.
     *
     * @param {string} src - The source directory.
     * @param {string} dest - The destination directory.
     * @param {function|null} onCopiedCallback - A callback function that is called after each file is copied.
     * @returns {string[]}
     */
    public recursiveDirectoryCopy(src: string, dest: string, onCopiedCallback: OnFileCopiedCallback | null = null) {
        const destFiles: string[] = [];
        const handler = (src, dest) => {
            const files = readdirSync(src);

            files
                .filter(file => file !== '.' && file !== '..')
                .forEach(file => {
                    const srcFn = src + '/' + file;
                    const destFn = dest + '/' + file;

                    const stats = lstatSync(srcFn, { throwIfNoEntry: false }) || null;

                    if (stats === null) {
                        return;
                    }

                    if (stats?.isFile() && !this.filesAreEqual(srcFn, destFn)) {
                        copyFileSync(srcFn, destFn);
                        destFiles.push(destFn);

                        if (onCopiedCallback) {
                            onCopiedCallback(srcFn, destFn);
                        }
                    }

                    if (stats?.isDirectory()) {
                        if (!this.fileexists(destFn)) {
                            mkdirSync(destFn);
                        }
                        handler(srcFn, destFn);
                    }
                });
            return destFiles;
        };

        return handler(src, dest);
    }
}
