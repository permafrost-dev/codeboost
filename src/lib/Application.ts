import { AppSettings, loadSettings } from '@/lib/AppSettings';
import { CodeBoost } from '@/lib/CodeBoost';
import { initOctokit } from '@/lib/github';
import { HistoryManager } from '@/lib/HistoryManager';
import { Repository } from '@/lib/Repository';

export class Application {
    public settings!: AppSettings;
    public codeboost!: CodeBoost;
    public repositoryName!: string;
    public historyManager!: HistoryManager;

    constructor() {
        this.historyManager = new HistoryManager(`${__dirname}/history.json`);
    }

    async init(boostId: string) {
        this.settings = loadSettings(`${__dirname}/settings.js`);
        initOctokit(this.settings.github_token);

        const repo = new Repository(this.repositoryName, this.settings.repository_storage_path);

        await repo.clone();

        if (this.settings.use_forks) {
            await repo.createFork();
        }

        this.codeboost = new CodeBoost(this.historyManager);
        await this.codeboost.init(boostId, ['8.2'], repo, this.settings);
    }

    async execute(repoName: string, boostName: string) {
        this.repositoryName = repoName;
        await this.init(boostName);
        this.historyManager.save();
    }
}
