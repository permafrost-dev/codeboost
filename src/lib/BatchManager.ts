import { Boost } from '@/lib/Boost';

export class BatchManager {
    public data: any[] = [];

    constructor(public filename: string) {
        this.data = require(filename);
    }

    public getBatch(boost: Boost, size: number) {
        //
    }

    protected getUsableDataset(boost: Boost) {
        return this.data.filter(item => boost.canRunOnRepository(item.repository));
    }

    protected getItemState(item: any) {
        return item.state || {};
    }
}
