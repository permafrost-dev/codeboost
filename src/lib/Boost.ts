import { AppSettings } from '@/lib/AppSettings';
import { BoostHistory, BoostHistoryItem, BoostHistoryItemState } from '@/types/BoostHistory';
import { CodeBoost } from '@/lib/CodeBoost';
import { generateRunId, versionToShortVersion } from '@/lib/helpers';
import { Repository } from '@/lib/Repository';
import { Tools } from '@/lib/Tools';
import { BoostConfiguration } from '@/types/BoostConfiguration';
import edge from 'edge.js';
import { existsSync, readFileSync } from 'fs';
import semver from 'semver';
import { SimpleGit } from 'simple-git';
import dayjs from 'dayjs';
import { Github } from '@/lib/github';

export interface BoostScriptHandlerParameters {
    /** arguments passed in from the user */
    args: any[];
    /** the boost instance */
    boost: Boost;
    /** information about the current boost run */
    currentRun: BoostHistoryItem;
    /** a simpleGit instance for the repository */
    git: SimpleGit;
    /** a collection of commonly-used libraries available to the boost scripts */
    libs: {
        fs: typeof import('fs');
        path: typeof import('path');
        semver: typeof import('semver');
    };
    /** the repository instance */
    repository: Repository;
    /** a collection of utility functions available to the boost scripts */
    tools: Tools;
}

export type BoostScriptHandler = (params: BoostScriptHandlerParameters) => any;

export class Boost {
    protected codeBoost!: CodeBoost;
    protected repository: Repository | null = null;

    public config!: BoostConfiguration;
    public path!: string;
    public id!: string;
    public version!: string;
    public repositoryLimits!: {
        maxRunsPerVersion: number;
        minutesBetweenRuns: number;
    };
    public pullRequest!: {
        title: string;
        body: string;
        branch: string;
    };
    public scripts!: any[];
    public actions!: any[];
    public state: Record<string, any> = {};
    public changedFiles: string[] = [];
    public runId: string;

    constructor(codeBoost: CodeBoost, boostPath: string) {
        this.runId = generateRunId();
        this.config = this.loadConfiguration(boostPath);
        this.codeBoost = codeBoost;
        this.init(boostPath);
    }

    public init(boostPath: string) {
        this.path = `${boostPath}/${this.config.version}`;
        this.id = this.config.id;
        this.version = this.config.version;

        this.repositoryLimits = {
            maxRunsPerVersion: this.config.repository_limits.max_runs_per_version,
            minutesBetweenRuns: this.config.repository_limits.minutes_between_runs,
        };

        this.pullRequest = this.loadPullRequest(this.config.pull_request);
        this.scripts = this.loadScripts(this.config.scripts.files);
    }

    public get appSettings(): AppSettings {
        return this.codeBoost.appSettings;
    }

    public get history(): BoostHistory {
        return this.codeBoost.historyManager.for(this.id);
    }

    public log(message) {
        this.codeBoost.log(message, [{ boost: this.id, run_id: this.runId, repository: this.repository?.fullRepositoryName() }]);
    }

    public loadConfiguration(boostPath: string) {
        const config = require(`${boostPath}/boost.js`).default;

        return <BoostConfiguration>config;
    }

    public loadPullRequest(pullRequest: Record<string, any>) {
        return {
            title: pullRequest.title,
            body: pullRequest.body,
            branch: `${pullRequest.branch}-v${versionToShortVersion(this.version)}`,
        };
    }

    public loadScripts(scripts: string[]) {
        const result: any[] = [];

        for (const script of scripts) {
            const fn = `${this.path}/${script}`;

            if (!existsSync(fn)) {
                throw new Error(`Boost script not found: ${fn}`);
            }

            result.push(require(fn).handler);
        }

        return result;
    }

    public async run(repository: Repository, args: any[] = []) {
        this.repository = repository;

        // changes to historyItem properties will be reflected in the saved history automatically
        // as createEntry returns a proxy object
        const historyItem: BoostHistoryItem = this.codeBoost.historyManager.createEntry({
            run_id: this.runId,
            boost: this.id,
            version: this.version,
            repository: repository.fullRepositoryName(),
            started_at: new Date().toISOString(),
            finished_at: null,
            state: BoostHistoryItemState.RUNNING,
            pull_request: null,
        });

        if (!this.canRunOnRepository(repository)) {
            historyItem.state = BoostHistoryItemState.SKIPPED;
            historyItem.finished_at = new Date().toISOString();

            this.log(`Cannot run on ${repository.fullRepositoryName()}`);
            this.log('Done.');
            return false;
        }

        const params = this.createScriptHandlerParameters(args, historyItem);

        const catchErrors = async (callBack: CallableFunction, args: any = []) => {
            try {
                return await callBack(...args);
            } catch (e) {
                console.log('******');
                hasError = true;
                this.log(e);
                return false;
            }
        };

        const handleDryRun = async (callback: CallableFunction, dryRunMessage: string) => {
            if (!this.appSettings.dry_run) {
                return await callback();
            }

            this.log(`[dry run] ${dryRunMessage}`);
            return null;
        };

        let hasError = false;

        if (!this.appSettings.use_pull_requests) {
            this.pullRequest.branch = await repository.defaultBranch();
        }

        if (this.appSettings.use_pull_requests) {
            await this.updatePullRequestBranchName();
            await this.checkoutPullRequestBranch();
        }

        catchErrors(async () => {
            await this.runInitializationScript(params);
            await this.runScripts(params);
        });

        if (!hasError && this.changedFiles.length > 0) {
            catchErrors(async () => {
                const remote = this.appSettings.use_forks ? 'fork' : 'origin';
                handleDryRun(
                    async () => await repository.git.push(remote, this.pullRequest.branch),
                    `push branch ${this.pullRequest.branch} to ${remote}`,
                );
            });

            catchErrors(async () => {
                handleDryRun(async () => {
                    await this.handlePullRequestCreation({ repository, historyItem });
                }, 'create pull request');
            });
        }

        historyItem.finished_at = new Date().toISOString();
        historyItem.state = hasError ? BoostHistoryItemState.FAILED : BoostHistoryItemState.SUCCEEDED;

        this.log('Done.');
    }

    public async handlePullRequestCreation({ repository, historyItem }) {
        if (!this.appSettings.use_pull_requests) {
            return;
        }

        const loadStringOrFile = (value: string) => {
            return existsSync(`${this.path}/${value}`) ? readFileSync(`${this.path}/${value}`, 'utf8') : value;
        };

        const title = await edge.renderRaw(loadStringOrFile(this.pullRequest.title), { boost: this, state: () => this.state });
        const body = await edge.renderRaw(loadStringOrFile(this.pullRequest.body), { boost: this, state: () => this.state });

        const defaultBranch = (await this.repository?.defaultBranch()) ?? 'main';

        const pr: any = await Github.createPullRequest(repository, this.pullRequest.branch, defaultBranch, title.trim(), body.trim());
        historyItem.pull_request = pr?.number;

        if (pr) {
            this.log(`created pull request #${pr.number}`);

            if (this.appSettings.auto_merge_pull_requests) {
                await Github.mergePullRequest(repository, pr.number);
                this.log(`merged pull request #${pr.number}`);
            }
        }
    }

    public createScriptHandlerParameters(args: any[], historyItem: BoostHistoryItem): BoostScriptHandlerParameters {
        return <BoostScriptHandlerParameters>{
            args,
            boost: this,
            currentRun: Object.freeze(Object.assign({}, historyItem)),
            git: this.repository?.git,
            libs: {
                fs: require('fs'),
                path: require('path'),
                semver,
            },
            repository: this.repository,
            tools: new Tools(),
        };
    }

    public async runInitializationScript(params: BoostScriptHandlerParameters) {
        if (existsSync(`${this.path}/init.js`)) {
            const initFn = require(`${this.path}/init.js`).handler;
            await initFn(params);
        }
    }

    public async checkoutPullRequestBranch() {
        await this.repository?.checkout(this.pullRequest.branch);
    }

    /**
     * update to a unique branch name if the branch already exists
     */
    public async updatePullRequestBranchName() {
        if (!this.repository) {
            return false;
        }

        const branches = await this.repository.localBranches();

        if (branches.all.includes(this.pullRequest.branch)) {
            let counter = 1;
            let newBranchName = this.pullRequest.branch;

            while (branches.all.includes(newBranchName)) {
                counter++;
                newBranchName = `${this.pullRequest.branch}-${counter}`;
            }

            this.pullRequest.branch = newBranchName;
        }

        return true;
    }

    public async runScripts(params: BoostScriptHandlerParameters) {
        if (this.config.scripts.parallel) {
            await Promise.allSettled(this.scripts.map(script => script(params)));
            return;
        }

        for (const script of this.scripts) {
            await script(params);
        }
    }

    public canRunOnRepository(repo: Repository | string) {
        const repoName = typeof repo === 'string' ? repo : repo.fullRepositoryName();

        const runs = this.history
            .filter(run => run.repository === repoName)
            .filter(run => run.version === this.version && run.run_id !== this.runId)
            .filter(run => run.state !== BoostHistoryItemState.SKIPPED);

        if (this.repositoryLimits.maxRunsPerVersion <= runs.length) {
            return false;
        }

        return !this.isRunTimeRestricted(runs);
    }

    protected isRunTimeRestricted(runs: BoostHistoryItem[]): boolean {
        for (const item of runs) {
            const runDate = dayjs(item.started_at);

            if (dayjs().diff(runDate, 'minute') < this.repositoryLimits.minutesBetweenRuns) {
                return true;
            }
        }

        return false;
    }
}
