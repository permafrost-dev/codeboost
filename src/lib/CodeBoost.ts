import { AppSettings } from '@/lib/AppSettings';
import { Boost } from '@/lib/Boost';
import { HasLogger, LogTarget } from '@/traits/HasLogger';
import { HistoryManager } from '@/lib/HistoryManager';
import { Repository } from '@/lib/Repository';
import { Mixin } from 'ts-mixer';

export class CodeBoost extends Mixin(HasLogger) {
    protected repository!: Repository;
    public appSettings!: AppSettings;
    public historyManager!: HistoryManager;
    public repositoryPrepared = false;

    constructor(appSettings: AppSettings, historyManager: HistoryManager) {
        super();
        this.createLogger(<LogTarget[]>appSettings?.log_target ?? [], {});
        this.appSettings = appSettings;
        this.historyManager = historyManager;
    }

    public async init(repository: Repository, appSettings: AppSettings) {
        this.appSettings = appSettings;
        this.repository = repository;
    }

    public async prepareRepository() {
        if (this.repositoryPrepared) {
            return;
        }

        await this.repository?.clone();
        await this.repository?.prepare();

        if (this.appSettings.use_forks && !this.appSettings.dry_run) {
            console.log('creating fork');
            await this.repository?.createFork();
        }

        this.repositoryPrepared = true;
    }

    public async runBoost(boost: string | Boost, args: string[]) {
        boost = typeof boost === 'string' ? this.getBoost(boost) : boost;

        await boost.run(this.repository, args);

        return boost;
    }

    public getBoost(boostName: string) {
        return new Boost(this, `${this.appSettings.boosts_path}/${boostName}`);
    }
}
