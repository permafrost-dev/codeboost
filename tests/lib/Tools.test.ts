import { Tools } from '@/lib/Tools';
import { execSync } from 'child_process';
import { existsSync, mkdirSync, readFileSync } from 'fs';

let tools: Tools;

beforeAll(() => {
    execSync(`rm -rf ${__dirname}/fixtures/temp/test*`);
});

beforeEach(() => {
    tools = new Tools();
});

it('sleeps for 1 second', async () => {
    const start = Date.now();
    await tools.sleep(1000);
    const end = Date.now();

    expect(end - start).toBeGreaterThanOrEqual(950);
});

it('checks if a file exists', () => {
    expect(tools.fileexists(__filename)).toBeTruthy();
    expect(tools.fileexists('non-existent-file')).toBeFalsy();
});

it('reads a file', () => {
    expect(tools.readfile(__filename)).toContain(`it('reads a file', () => {`);
});

it('writes a file', () => {
    const path = `${__dirname}/../fixtures/temp/test.txt`;
    const content = 'Hello, world!';

    tools.writefile(path, content);

    expect(existsSync(path)).toBeTruthy();
    expect(readFileSync(path, 'utf-8')).toBe(content);
});

it('copies a file', () => {
    const src = __filename;
    const dest = `${__dirname}/../fixtures/temp/test.ts`;

    expect(existsSync(dest)).toBeFalsy();

    tools.copyfile(src, dest);

    expect(existsSync(dest)).toBeTruthy();
});

it('hashes a file', () => {
    const path = `${__dirname}/../fixtures/test-boost-1/1.0.0/first.js`;

    expect(tools.hashfile(path)).toBe('ebfa68fa54781a061bc26dde44e10a82df2bc1ded4aff93df82a505f364eb553');
});

it('hashes a string', () => {
    const str = 'test string';

    expect(tools.hashstring(str)).toBe('d5579c46dfcc7f18207013e65b44e4cb4e2c2298f4ac457ba8f82743f31e930b');
});

it('compares two files', () => {
    const path1 = `${__dirname}/../fixtures/test-boost-1/1.0.0/first.js`;
    const path2 = `${__dirname}/../fixtures/test-boost-1/1.0.0/second.js`;

    expect(tools.filesAreEqual(path1, path2)).toBeTruthy();
});

it('reads a yaml file', () => {
    const path = `${__dirname}/../fixtures/files/test1.yaml`;

    expect(tools.readYaml(path)).toMatchSnapshot();
});

it('reads a json file', () => {
    const path = `${__dirname}/../fixtures/files/test1.json`;

    expect(tools.readJson(path)).toMatchSnapshot();
});

it('copies a directory recursively', () => {
    const src = `${__dirname}/../fixtures/test-boost-1`;
    const dest = `${__dirname}/../fixtures/temp/test-boost-1`;

    mkdirSync(dest, { recursive: true });

    tools.recursiveDirectoryCopy(src, dest);

    expect(existsSync(dest)).toBeTruthy();
    expect(tools.filesAreEqual(`${src}/1.0.0/first.js`, `${dest}/1.0.0/first.js`)).toBeTruthy();
});
