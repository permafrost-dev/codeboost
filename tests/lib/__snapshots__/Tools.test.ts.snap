// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reads a json file 1`] = `
{
  "one": 1,
}
`;

exports[`reads a yaml file 1`] = `
{
  "one": 1,
}
`;
